File: bindings/ruby/lib/proven/safe_hex.rb
# SPDX-License-Identifier: PMPL-1.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
# frozen_string_literal: true

# Safe hexadecimal operations via FFI to libproven.
# ALL computation delegates to Idris 2 compiled code.

module Proven
  module SafeHex
    class << self
      # Encode bytes to lowercase hex string.
      # Returns nil on error.
      #
      # @param bytes [String] binary data
      # @return [String, nil]
      def encode(bytes)
        return nil if bytes.nil?
        ptr, len = FFI.str_to_ptr(bytes)
        result = FFI.invoke_string_result(
          "proven_hex_encode",
          [Fiddle::TYPE_VOIDP, Fiddle::TYPE_SIZE_T, Fiddle::TYPE_INT],
          [ptr, len, 0] # uppercase = false
        )
        return nil unless result
        status, str = result
        status == FFI::STATUS_OK ? str : nil
      end

      # Encode bytes to uppercase hex string.
      # Returns nil on error.
      #
      # @param bytes [String] binary data
      # @return [String, nil]
      def encode_upper(bytes)
        return nil if bytes.nil?
        ptr, len = FFI.str_to_ptr(bytes)
        result = FFI.invoke_string_result(
          "proven_hex_encode",
          [Fiddle::TYPE_VOIDP, Fiddle::TYPE_SIZE_T, Fiddle::TYPE_INT],
          [ptr, len, 1] # uppercase = true
        )
        return nil unless result
        status, str = result
        status == FFI::STATUS_OK ? str : nil
      end

      # Decode hex string to binary bytes.
      # Returns nil on invalid hex input.
      #
      # @param hex_string [String]
      # @return [String, nil] binary string
      def decode(hex_string)
        return nil if hex_string.nil?
        ptr, len = FFI.str_to_ptr(hex_string)

        # HexDecodeResult = { i32 status, ptr data, size_t length }
        # Same layout as StringResult (24 bytes with sret)
        buf = Fiddle::Pointer.malloc(32, Fiddle::RUBY_FREE)

        fn = Fiddle::Function.new(
          FFI.handler["proven_hex_decode"],
          [Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP, Fiddle::TYPE_SIZE_T],
          Fiddle::TYPE_VOID,
          need_gvl: true
        )
        fn.call(buf, ptr, len)

        packed = buf.to_str(24)
        status = packed[0, 4].unpack1("l!")
        return nil unless status == FFI::STATUS_OK

        data_ptr_val = packed[8, 8].unpack1("Q")
        length = packed[16, 8].unpack1("Q")

        if data_ptr_val != 0 && length > 0
          data_ptr = Fiddle::Pointer.new(data_ptr_val)
          bytes = data_ptr.to_str(length).dup

          # Free via proven_hex_free
          begin
            free_fn = Fiddle::Function.new(
              FFI.handler["proven_hex_free"],
              [Fiddle::TYPE_VOIDP],
              Fiddle::TYPE_VOID
            )
            free_fn.call(buf)
          rescue Fiddle::DLError
            # If free not available, best-effort
          end

          bytes
        else
          nil
        end
      rescue Fiddle::DLError, TypeError
        nil
      end
    end
  end
end

File: bindings/php/src/Proven/SafeHex.php
<?php
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
//
// SafeHex - FFI wrapper for proven_hex_* functions.
// All computation delegated to libproven; no logic reimplemented here.

declare(strict_types=1);

namespace Proven;

/**
 * Safe hexadecimal encoding/decoding via libproven FFI.
 *
 * Every method calls the corresponding libproven FFI function.
 */
final class SafeHex
{
    /**
     * Encode raw bytes to hexadecimal string.
     *
     * @param string $data      Raw bytes to encode.
     * @param bool   $uppercase If true, use uppercase hex digits.
     * @return string|null The hex-encoded string, or null on error.
     */
    public static function encode(string $data, bool $uppercase = false): ?string
    {
        return self::extractString(
            FFI::getLib()->proven_hex_encode($data, strlen($data), $uppercase)
        );
    }

    /**
     * Decode a hexadecimal string to raw bytes.
     *
     * @param string $hex The hex string to decode.
     * @return string|null The decoded bytes, or null on error.
     */
    public static function decode(string $hex): ?string
    {
        $result = FFI::getLib()->proven_hex_decode($hex, strlen($hex));
        if ($result->status !== 0 || $result->data === null) {
            return null;
        }
        $bytes = \FFI::string($result->data, $result->length);
        FFI::getLib()->proven_hex_free(\FFI::addr($result));
        return $bytes;
    }

    /**
     * Extract string from StringResult, free C memory, return null on error.
     */
    private static function extractString(mixed $result): ?string
    {
        if ($result->status !== 0 || $result->value === null) {
            return null;
        }
        $str = \FFI::string($result->value, $result->length);
        FFI::getLib()->proven_free_string($result->value);
        return $str;
    }
}
